filters/imagefilter.go

```diff
@@ -43,65 +41,74 @@ func init() {
 
 type ImageFilter interface {
 	CreateOptions(image *bimg.Image) (*bimg.Options, error)
+	CanBeMerged(other *bimg.Options, self *bimg.Options) bool
+	Merge(other *bimg.Options, self *bimg.Options) *bimg.Options
 }
 
 func HandleImageResponse(ctx filters.FilterContext, f ImageFilter) {
-	rsp := ctx.Response()
-
-	defer rsp.Body.Close()
-
-	rsp.Header.Del("Content-Length")
-
-	r, err := handleImageTransform(rsp.Body, f)
+	image := ctx.StateBag()[SkropImage].(*bimg.Image)
 
+	opt, err := f.CreateOptions(image)
 	if err != nil {
-		log.Error("failed to process image ", err.Error())
+		log.Error("Failed to create options ", err.Error())
 		ctx.Serve(&http.Response{
-			StatusCode:http.StatusInternalServerError,
-			Body: ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
 		})
 		return
 	}
 
-	rsp.Body = ioutil.NopCloser(r)
-}
+	opts := ctx.StateBag()[SkropOptions].(*bimg.Options)
 
-func handleImageTransform(r io.Reader, f ImageFilter) (io.Reader, error) {
-	var err error
-	imageBytes, err := ioutil.ReadAll(r)
+	if f.CanBeMerged(opts, opt) {
+		ctx.StateBag()[SkropOptions] = f.Merge(opts, opt)
+		log.Debug("Filter ", f, " merged in ", ctx.StateBag()[SkropOptions])
+		return
+	}
 
+	//transform the image
+	buf, err := transformImage(image, opts)
```

Comment:
as discussed offline, there is a major problem: some filters return a bimg.Option based on the image size
example:
original image is Width=1000 Height=500
filter1 is crop(300, 150)
filter2 is resizeLongerEdge(100)
filter2 checks which one is the longer edge and resize it to 100. Because the filter2 cannot be merged with the filter1 if I process the image, the second filter will apply the 100 on the Width, if I don't process the image the filter2 will apply the resize on the Height